docs/concepts/time-and-windows-in-ksql-queries.rst

```diff
@@ -0,0 +1,305 @@
+time of processing to implement processing-time semantics.
+
+By creating a custom timestamp extractor class, you can enforce different notions
+or semantics of time, depending on the requirements of your business logic. For
+more information see :ref:`streams_developer-guide_timestamp-extractor`.
+
+Windows in KSQL Queries
+***********************
+
+Representing time consistently enables aggregation operations on streams and
+tables, like SUM, that have distinct time boundaries. In KSQL, these boundaries
+are named *time windows*.
+
+.. image:: ../img/ksql-window.png
+   :alt: Diagram showing the relationship between events and time in a KSQL stream
+
+Windowing lets you control how to group records that have the same key for
+stateful operations, like aggregations or joins, into time spans. KSQL tracks
+windows per record key.
+
+When using windows in your KSQL queries, aggregate functions are applied only
+to the records that occur within a specific time window. Records that arrive
+late are handled as you might expect: although the time window they belong to
+has expired, the late records are still associated with the correct window.
+
+You can specify a retention period for the window in your KSQL queries. This
+retention period controls how long KSQL waits for out-of-order or late-arriving
+records for a given window. If a record arrives after the retention period of a
+window has passed, the record is discarded and isn't processed in that window.
+
+.. Note for version 5.1
+    Starting in KSQL 5.1, a *grace period* configuration determines how long
+    to wait before closing a window. Retention time is still a valid parameter
+    that defines for how long the potentially closed window is stored. This
+    enables accessing the window with interactive queries, even if it's closed.
+
+In the real world, late-arriving records are always possible, and your KSQL
+applications must account for them properly. The system's time semantics
+determine how late records are handled. For processing-time, the semantics are
+“when the record is being processed”, which means that the notion of late records
+isn't applicable because, by definition, no record can be late.
+
+Late-arriving records are considered “late” only for event-time and ingestion-time
+semantics. In both cases, KSQL is able to handle late-arriving records properly.
+
+.. important::
+
+    KSQL is based on the Unix epoch time in the UTC timezone, and this can affect
+    time windows. For example, if you define a 24-hour tumbling time window, it will
+    be in the UTC timezone, which may not be appropriate if you want to have daily
+    windows in your timezone.
+
+Window Types
+============
+
+There are three ways to define time windows in KSQL.
+
+.. image:: ../img/ksql-window-aggregation.png
+   :alt: Diagram showing three types of time windows in KSQL streams: tumbling, hopping, and session
+
+Tumbling Window

```

Comment:
We have a diagram for tumbling windows in the KStreams docs (https://docs.confluent.io/current/streams/developer-guide/dsl-api.html#tumbling-time-windows). Don't we want to re-use that here to help users understand tumbling windows more easily?